src/mesh.cpp

```diff
@@ -0,0 +1,981 @@
+  // (x,z)
+  if ((y0 < ym0 && y1 > ym0) || (y0 > ym0 && y1 < ym0)) {
+    double xi = x0 + (ym0 - y0) * (x1 - x0) / (y1 - y0);
+    double zi = z0 + (ym0 - y0) * (z1 - z0) / (y1 - y0);
+    if (xi >= xm0 && xi < xm1 && zi >= zm0 && zi < zm1) {
+      return true;
+    }
+  }
+
+  // Check if line intersects bottom surface -- calculate the intersection
+  // point (x,y)
+  if ((z0 < zm0 && z1 > zm0) || (z0 > zm0 && z1 < zm0)) {
+    double xi = x0 + (zm0 - z0) * (x1 - x0) / (z1 - z0);
+    double yi = y0 + (zm0 - z0) * (y1 - y0) / (z1 - z0);
+    if (xi >= xm0 && xi < xm1 && yi >= ym0 && yi < ym1) {
+      return true;
+    }
+  }
+
+  // Check if line intersects right surface -- calculate the intersection point
+  // (y,z)
+  if ((x0 < xm1 && x1 > xm1) || (x0 > xm1 && x1 < xm1)) {
+    double yi = y0 + (xm1 - x0) * (y1 - y0) / (x1 - x0);
+    double zi = z0 + (xm1 - x0) * (z1 - z0) / (x1 - x0);
+    if (yi >= ym0 && yi < ym1 && zi >= zm0 && zi < zm1) {
+      return true;
+    }
+  }
+
+  // Check if line intersects front surface -- calculate the intersection point
+  // (x,z)
+  if ((y0 < ym1 && y1 > ym1) || (y0 > ym1 && y1 < ym1)) {
+    double xi = x0 + (ym1 - y0) * (x1 - x0) / (y1 - y0);
+    double zi = z0 + (ym1 - y0) * (z1 - z0) / (y1 - y0);
+    if (xi >= xm0 && xi < xm1 && zi >= zm0 && zi < zm1) {
+      return true;
+    }
+  }
+
+  // Check if line intersects top surface -- calculate the intersection point
+  // (x,y)
+  if ((z0 < zm1 && z1 > zm1) || (z0 > zm1 && z1 < zm1)) {
+    double xi = x0 + (zm1 - z0) * (x1 - x0) / (z1 - z0);
+    double yi = y0 + (zm1 - z0) * (y1 - y0) / (z1 - z0);
+    if (xi >= xm0 && xi < xm1 && yi >= ym0 && yi < ym1) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void RegularMesh::bins_crossed(const Particle* p, std::vector<int>& bins,
+                               std::vector<double>& lengths)
+{
+  constexpr int MAX_SEARCH_ITER = 100;
+
+  // ========================================================================
+  // Determine if the track intersects the tally mesh.
+
+  // Copy the starting and ending coordinates of the particle.  Offset these
+  // just a bit for the purposes of determining if there was an intersection
+  // in case the mesh surfaces coincide with lattice/geometric surfaces which
+  // might produce finite-precision errors.
+  Position last_r {p->last_xyz};
+  Position r {p->coord[0].xyz};
+  Direction u {p->coord[0].uvw};
+
+  Position r0 = last_r + TINY_BIT*u;
+  Position r1 = r - TINY_BIT*u;
+
+  // Determine indices for starting and ending location.
+  int n = n_dimension_;
+  xt::xtensor<int, 1> ijk0 = xt::empty<int>({n});
+  bool start_in_mesh;
+  get_indices(r0, ijk0.data(), &start_in_mesh);
+  xt::xtensor<int, 1> ijk1 = xt::empty<int>({n});
+  bool end_in_mesh;
+  get_indices(r1, ijk1.data(), &end_in_mesh);
+
+  // If this is the first iteration of the filter loop, check if the track
+  // intersects any part of the mesh.
```

Comment:
Remove the `If this ... filter loop` part of the comment.  That's a holdover from when filters only provided bins one-at-a-time.